src/classes/HiFiAudioAPIData.ts

```diff
@@ -66,33 +101,249 @@ export class OrientationEuler3D {
+    case OrientationEuler3DOrder.YawRollPitch: {
+        return new OrientationQuat3D({
+                x: ax + bx,
+                y: ay + by,
+                z: az - bz,
+                w: aw - bw,
+            });
+        } break;
+  
+    // From 'base' space rotate Pitch, then Roll then Yaw...
+    case OrientationEuler3DOrder.PitchRollYaw: {
+        return new OrientationQuat3D({
+                x: ax - bx,
+                y: ay - by,
+                z: az + bz,
+                w: aw + bw,
+            });
+        } break;
+    }    
+}
+
+/**
+ * Compute the orientation euler decomposition from the specified quaternion.
+ * The resulting euler is the rotation transforming from combining the euler angles rotations in the specified order
+ * 
+ * For example, the order YawPitchRoll is computed as follow:
+ *  starting from the base 3d frame,
+ *  1/ Yaw, rotating around the vertical axis
+ *  2/ Pitch, rotating around the right axis 
+ *  3/ Roll, rotating around the front axis
+ *  the resulting 3d frame orientation is relative to the base frame.
+ *  The resulting rotation is defining the 'rotated' space relative to the 'base' space.
+ *  A vector Vr in "rotated' space and its equivalent value Vb in the'base' space is computed as follow:
+ *  Vb = [P][Y][R] Vr
+ * 
+ * @param quat - The orientation quaternion.
+ * @param order - The euler order convention.
+ * 
+ * @return The end resulting quaternion defined from the euler angles combination
+ */
+export function eulerFromQuaternion(quat: OrientationQuat3D, order : OrientationEuler3DOrder): OrientationEuler3D {
+    // We need to convert the quaternion to the equivalent mat3x3
+    let qx2 = quat.x * quat.x;
+    let qy2 = quat.y * quat.y;
+    let qz2 = quat.z * quat.z;
+    // let qw2 = quat.w * quat.w; we could choose to use it instead of the 1 - 2* term...
+    let qwx = quat.w * quat.x;
+    let qwy = quat.w * quat.y;
+    let qwz = quat.w * quat.z;
+    let qxy = quat.x * quat.y;
+    let qyz = quat.y * quat.z;
+    let qxz = quat.z * quat.x;
+    // ROT Mat33 =  {  1 - 2qy2 - 2qz2  |  2(qxy - qwz)    |  2(qxz + qwy)  }
+    //              {  2(qxy + qwz)     |  1 - 2qx2 - 2qz2 |  2(qyz - qwx)  }
+    //              {  2(qxz - qwy)     |  2(qyz + qwx)    |  1 - 2qx2 - 2qy2  }
+    let r00 = 1.0 - 2.0 * (qy2 + qz2);
+    let r10 = 2.0 * (qxy + qwz);
+    let r20 = 2.0 * (qxz - qwy);
+
+    let r01 = 2.0 * (qxy - qwz);
+    let r11 = 1.0 - 2.0 * (qx2 + qz2); 
+    let r21 = 2.0 * (qyz + qwx);
+   
+    let r02 = 2.0 * (qxz + qwy);
+    let r12 = 2.0 * (qyz - qwx);
+    let r22 = 1.0 - 2.0 * (qx2 + qy2); 
+
+    // then depending on the euler rotation order decomposition, we extract the angles 
+    // from the base vector components
+    let pitch = 0;
+    let yaw = 0;
+    let roll = 0;
+    switch (order) {
+    case OrientationEuler3DOrder.PitchYawRoll: {
+        yaw = Math.asin( r02 );
+        if ( Math.abs( r02 ) < 0.9999999 ) {
```

Comment:
any reason not to do this here and below?:

```suggestion
        if ( Math.abs( r02 ) < 1 - Number.EPSILON ) {
```